### PR TITLE
add rolling window batch operation for tf.data.Dataset 

### LINKED ISSUE #15044: rolling window batch operation for tf.data.Dataset
_This is a feature request_

For Datasets that represent a sequence or time series, it can be useful to have a Dataset op that creates a rolling window batch over the given Dataset. For example, if I have a tf.data.Dataset whose elements represent a time series (line breaks separate elements):
```
1
2
3
4
5
6
7
8
9
```
The rolling window batch would create a dataset with the following elements (for window size 4 and stride 1):
```
1 2 3 4
2 3 4 5
3 4 5 6
4 5 6 7
5 6 7 8
6 7 8 9
```
This operation will be extremely useful for extracting sub sequences from a time series for training RNNs and Reinforcement Learning models.

### PR DESCRIPTION
**Pull request by facaiy**
Resolve #15044.

### implementation

The PR proposes a `slide` method for Dataset: Groups elements in fixed size blocks by passing a "sliding window" over Dataset. It behaves like `batch`, in fact, `batch(n) == slide(n, n)`.

~I failed to move c++ implementation from `core` to `contrib`. Any help will be appreciated.~

### how to test

+ [x] add test case.
+ [x] pass all tests.

### CODE REVIEW

File: tensorflow/contrib/data/python/ops/sliding.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  def output_shapes(self):
+    input_shapes = self._input_dataset.output_shapes
+    return nest.pack_sequence_as(input_shapes, [
+      tensor_shape.vector(None).concatenate(s)
+      for s in nest.flatten(self._input_dataset.output_shapes)
+    ])
+
+  @property
+  def output_types(self):
+    return self._input_dataset.output_types
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/sliding.py]: Can you add a function to this file that creates a `SlideDataset` can be used with `tf.data.Dataset.apply()`? I'd be inclined to give it a slightly more verbose name than "slide", e.g. `sliding_window_batch():`.

See e.g. the implementation of `tf.contrib.data.map_and_batch()`:

https://github.com/tensorflow/tensorflow/blob/23ed83de415b465fb5570216b8c171d63187289e/tensorflow/contrib/data/python/ops/batching.py#L387

File: tensorflow/contrib/data/python/ops/dataset_ops.py [hunk 2]
```diff
@@ -428,6 +429,21 @@ def batch(self, batch_size):
...
+        elements in the window.
+      slide_step: (Optional.) A `tf.int64` scalar `tf.Tensor`, representing
+        the steps moving forward for one iteration. The default is `1`. It
+        must be in [1, `slide_size`).
+
+    Returns:
+      A `Dataset`.
+    """
+    return Dataset(sliding.SlideDataset(self, slide_size, slide_step))
+
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/dataset_ops.py]: Please revert this file. The `tf.contrib.data.Dataset` class is deprecated, and we don't want to encourage any new users :). Using `Dataset.apply()` together with a `tf.contrib.data` function defined in `sliding.py` should suffice.

File: tensorflow/core/ops/dataset_ops.cc [hunk 3]
```diff
@@ -399,6 +399,22 @@ batch_size: A scalar representing the number of elements to accumulate in a
   batch.
 )doc");
 
+REGISTER_OP("SlideDataset")
+    .Input("input_dataset: variant")
+    .Input("slide_size: int64")
```
Comment (mrry) [inline tensorflow/core/ops/dataset_ops.cc]: Naming nit: I think `window_size` would be less ambiguous. (Is the "slide size" the size of the window or the size of the slide from one window to the next?)

File: tensorflow/core/ops/dataset_ops.cc [hunk 4]
```diff
@@ -399,6 +399,22 @@ batch_size: A scalar representing the number of elements to accumulate in a
...
+    .Input("slide_step: int64")
```
Comment (mrry) [inline tensorflow/core/ops/dataset_ops.cc]: Naming nit: I'd prefer `stride` for this argument name.

File: tensorflow/core/ops/dataset_ops.cc [hunk 5]
```diff
@@ -399,6 +399,22 @@ batch_size: A scalar representing the number of elements to accumulate in a
...
+    .Output("handle: variant")
+    .Attr("output_types: list(type) >= 1")
+    .Attr("output_shapes: list(shape) >= 1")
+    .SetShapeFn(shape_inference::ScalarShape)
+    .Doc(R"doc(
+Creates a dataset that slides a fixed window on `input_dataset`.
+
+slide_size: A scalar representing the number of elements in the fixed window.
+slide_step: A scalar representing the steps moving forward in one iteration.
+  The default is `1`. It must be in [1, `slide_size`).
```
Comment (mrry) [inline tensorflow/core/ops/dataset_ops.cc]: Note that there are no defaults at this level of the API (for tensor-valued inputs), so please delete this line.

File: tensorflow/core/kernels/data/slide_dataset_op.cc [hunk 6]
```diff
@@ -0,0 +1,253 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Comment (mrry) [inline tensorflow/core/kernels/data/slide_dataset_op.cc]: Please update all the copyright dates to 2018.

File: tensorflow/core/kernels/data/slide_dataset_op.cc [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+          }
+          // Fill up with new elements.
+          *end_of_sequence = false;
+          for (size_t i = batch_elements.size(); i < slide_size && !*end_of_sequence;
+              ++i) {
+            std::vector<Tensor> batch_element_tuple;
+            TF_RETURN_IF_ERROR(input_impl_->GetNext(ctx, &batch_element_tuple,
+                                                    end_of_sequence));
+            if (!*end_of_sequence) {
+              batch_elements.emplace_back(std::move(batch_element_tuple));
```
Comment (mrry) [inline tensorflow/core/kernels/data/slide_dataset_op.cc]: Nit: I think `push_back(std::vector<Tensor>&&)` would be preferable here.

File: tensorflow/core/kernels/data/slide_dataset_op.cc [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            } else {
+              input_impl_.reset();
+            }
+          }
+
+          if (batch_elements.size() < slide_size) {
```
Comment (mrry) [inline tensorflow/core/kernels/data/slide_dataset_op.cc]: Does this operation intentionally drop the last window if it's smaller than `slide_size`? Can you please add something about that to the documentation if so?

File: tensorflow/python/ops/hidden_ops.txt [hunk 9]
```diff
@@ -387,3 +387,6 @@ Fact
 # word2vec deprecated ops
 NegTrain
 Skipgram
+
+# dataset_ops
+SlideDataset
```
Comment (mrry) [inline tensorflow/python/ops/hidden_ops.txt]: I think you can revert this change... none of the other Dataset ops are in here.

**Review by mrry — CHANGES_REQUESTED** [review]
Thanks for the contribution! I have a few comments, mostly about naming, but this will be a welcome addition to the API.

(Thanks for trying to put the C++ code in contrib/... this doesn't currently work, but we're trying to make it possible, and might move the code at a later point.)

File: tensorflow/core/kernels/data/slide_dataset_op.cc (hunk 8 above)
Comment (carlthome) [reply tensorflow/core/kernels/data/slide_dataset_op.cc]: And if so, please add optional padding like in `tf.contrib.signal.frame`!

Comment (facaiy) [reply tensorflow/core/kernels/data/slide_dataset_op.cc]: Yes, it's intentional. Say, 
```python
dataset = [1, 2, 3, 4, 5]
slide_size=3
slide_step=1
```
In most case, I think we expect that:
```
[1, 2, 3]
     [2, 3, 4]
          [3, 4, 5]
```
So I'd prefer to drop it. And padding looks a little counterintuitive. I'd like to explain the behavior in the documentation later.

Comment (facaiy) [reply tensorflow/core/kernels/data/slide_dataset_op.cc]: Because dataset supports composite structure, it seems not easy for padding, say:
```
[ [1, 2], [[3.0, 4.0, 5.0]], [[[1+2j]]], ["hello"] ]
```

**Comment by facaiy** [comment]
Thanks for review, I have resolve all the comments except of padding, because I have no good idea of how to do it.

**Comment by facaiy** [comment]
Moreover, I find it ambiguous for padding.
For example:
```python
dataset = [1, 2, 3, 4, 5, 6, 7]
window_size = 4
stride = 2
padding_values = 0
```
Which one of the behaviors below do we expect?
+ a:
```
[1, 2, 3, 4]
    [3, 4, 5, 6]
         [5, 6, 7, 0]
```
+ b:
```
[1, 2, 3, 4]
    [3, 4, 5, 6]
         [5, 6, 7, 0]
             [7, 0, 0, 0]
```

**Comment by facaiy** [comment]
I have checked those failed logs: build broken. And those tests run well on my work machine. Seems unrelated?

**Comment by facaiy** [comment]
Thanks, I'll try to fix the problem when I get back from holiday.

```CPP
tensorflow/core/kernels/data/slide_dataset_op.cc:208:14: error: 'tensorflow::Status tensorflow::{anonymous}::SlideDatasetOp::Dataset::Iterator::RestoreInternal(tensorflow::OpKernelContext*, tensorflow::IteratorStateReader*)' marked override, but does not override
       Status RestoreInternal(OpKernelContext* ctx,
              ^
In file included from ./tensorflow/core/framework/tensor_shape.h:23:0,
                 from ./tensorflow/core/framework/partial_tensor_shape.h:20,
                 from tensorflow/core/kernels/data/slide_dataset_op.cc:15:
tensorflow/core/kernels/data/slide_dataset_op.cc: In member function 'tensorflow::Status tensorflow::{anonymous}::SlideDatasetOp::Dataset::Iterator::RestoreInternal(tensorflow::OpKernelContext*, tensorflow::IteratorStateReader*)':
tensorflow/core/kernels/data/slide_dataset_op.cc:212:68: error: no matching function for call to 'tensorflow::{anonymous}::SlideDatasetOp::Dataset::Iterator::RestoreParent(tensorflow::OpKernelContext*&, tensorflow::IteratorStateReader*&, std::unique_ptr<tensorflow::IteratorBase>&)'
           TF_RETURN_IF_ERROR(RestoreParent(ctx, reader, input_impl_));
                                                                    ^
```

**Comment by facaiy** [comment]
Hi, I have merged the latest codes from upstream, and fixed the broken test (at least for Linux CPU). Dose anyone help run all tests again? Thanks.

**Comment by facaiy** [comment]
@drpngx Hi, could you help me run all tests again? 7 days passed, and reviewers seem so busy to reply.

File: tensorflow/contrib/data/python/ops/sliding.py (hunk 1 above)
Comment (drpngx) [reply tensorflow/contrib/data/python/ops/sliding.py]: @facaiy did you address this comment? Thank you.

Comment (facaiy) [reply tensorflow/contrib/data/python/ops/sliding.py]: @drpngx Yes, `sliding_window_batch` has been added as suggested. Some doc build broken, and I fixed the failed tests. Could you help restart all tests? Thank you.

**Comment by facaiy** [comment]
Thank @drpngx for your help, all tests passed.

@mrry @rmlarsen I think all comments have been addressed, and could you take a look? Thank you.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/data/python/ops/sliding.py (hunk 1 above)
Comment (mrry) [reply tensorflow/contrib/data/python/ops/sliding.py]: Sorry for the delay... this comment is addressed now, thanks!

**Review by mrry — APPROVED** [review]
Thanks, and sorry again for the delay!

**Comment by facaiy** [comment]
Thanks all for the help!

**Comment by drpngx** [comment]
Woohoo!

**Comment by rombin** [comment]
Hi, I have issues getting this working

AttributeError: module 'tensorflow.python.ops.gen_dataset_ops' has no attribute 'slide_dataset'

any suggestions greatly appreciated, thank you

**Comment by facaiy** [comment]
@rombin Could you collect all necessary information and open a issue to report the problem?

**Comment by rombin** [comment]
I think I have an issue with installation. I have installed tensorflow 1.6 via Conda. But I do not see the sliding.py file in the contrib/data/python/ops folder. Do I somehow need even newer version? I notice you merged 2 days ago??

**Comment by facaiy** [comment]
Yes, I think the op will be released in the next version, and you can also install [tf-nightly](https://github.com/tensorflow/tensorflow#installation) to take a try.

**Comment by rombin** [comment]
tried tf-nightly, works like charm, thank you
